Runme Widget

christian-bromann opened the pull request:
This pull request implements a new widget to promote [Runme](https://runme.dev/). It comes with the following states:

### User has Runme not installed

![Screenshot 2023-03-06 at 18 54 44](https://user-images.githubusercontent.com/731337/223308298-8f262026-2bb3-4ec9-be7f-3310a696c9bc.png)

Clicking on `Add to VS Code` will open this link `vscode://stateful.runme?command=setup&fileToOpen=https://gist.githubusercontent.com/christian-bromann/df97ce3dace21564ffdf1900400ec099/raw/0a9b29c979d61e17032c855b04dbc4f5b962f847/ThankYou.md` which basically will prompt the install procedure for Run with Runme.

### No Runme Files are found

![createrunme](https://user-images.githubusercontent.com/731337/223308844-d1049bc6-6445-4303-a8d5-fc1da5832f94.gif)

If no markdown files are found in the workspace, the widget will ask whether the user wants to create one. If the link is click, it runs the Runme command `runme.new`.

### Open Runme Files

![openrunme](https://user-images.githubusercontent.com/731337/223309174-4b488aff-9730-466d-b009-8a2fcd4211c8.gif)

On click on the list item, the widget will run the `runme.openAsRunmeNotebook` command. There is a file listener registered so every markdown file added to the workspace or removed will have an impact on the list and automatically updates it.

---

@sourishkrout @admc please let me know if the widget fills the requirements. If functionality wise we are happy with the state I can wrap up the work adding unit/e2e tests.

admc commented:
Looking good! Couple thoughts:
- The runme files list uses quite a lot of space, i'd probably want it condensed to one line each, maybe small runme icon, filename and then a truncated grayed out path, also maybe grouped if there are many in one directory?
- After installing Runme, do we think people will get "Runme Notebooks" or should we use a title thats more literal like "Runnable Markdowns" or something? Just a thought.
- "Download for integrated Marquee support" makes me think of leaving VS Code and downloading something, maybe "Install native VS Code support for Runme" or something?

Otherwise the workflow seems right on. 👍

christian-bromann commented:
> * The runme files list uses quite a lot of space, i'd probably want it condensed to one line each, maybe small runme icon, filename and then a truncated grayed out path, also maybe grouped if there are many in one directory?

I condensed it a bit. I would still keep it on two lines because it better guarantees that users can see the full directory path. I was looking into grouping by directory but it is generally less likely that many markdowns are in the same directory. What do you think about this version:

![listcompact](https://user-images.githubusercontent.com/731337/223571896-2a39850b-a232-4872-8c01-325fc3189b1b.gif)

> * Runnable Markdowns

Modified widget title.

> * "Download for integrated Marquee support" makes me think of leaving VS Code and downloading something, maybe "Install native VS Code support for Runme" or something?

Updated:
<img width="669" alt="Screenshot 2023-03-07 at 14 47 44" src="https://user-images.githubusercontent.com/731337/223572027-ca2b2961-66b6-462b-9f73-58edca5e4359.png">

admc commented:
LGTM 👍

sourishkrout commented on packages/extension/src/gui.view.ts:
```diff
@@ -219,6 +219,8 @@ export class MarqueeGui extends EventEmitter {
 
     const backendBaseUrl = vscode.Uri.parse(BACKEND_BASE_URL)
     const backendGeoUrl = vscode.Uri.parse(BACKEND_GEO_URL)
+    console.log('OPEN WITH', widgetStateConfigurations)
```
Leftover debug statement?

sourishkrout commented on packages/extension/src/gui.view.ts:
```diff
@@ -299,6 +301,10 @@ export class MarqueeGui extends EventEmitter {
     }
 
+    if (args && args.length > 0 && command === 'runme.openAsRunmeNotebook') {
+      return vscode.commands.executeCommand(command, vscode.Uri.file(args[0].replace('file://', '')), options)
```
let's be sure this works on Windows

christian-bromann replied:
tested in on Windows, found some minor details but overall it worked.

sourishkrout reviewed (approved):
👍 LGTM
